_posts/2018-04-13-preliminary-sandboxfs-support.md

```diff
@@ -0,0 +1,168 @@
+---
+layout: posts
+title: Preliminary sandboxfs support and performance results
+---
+
+[Back in August of 2017]({% post_url 2017-08-25-introducing-sandboxfs %}), we
+introduced sandboxfs: a project to improve the performance and correctness of
+builds that have action sandboxing enabled. Today, after months of work to
+stabilize the codebase, we are happy to announce that preliminary support for
+sandboxfs is available in Bazel 0.12.0!
+
+This post presents the performance measurements we have gotten so far when using
+sandboxfs. As these metrics look promising, the true goal of this post is a
+**call for action**: we know a bunch of you have previously expressed that
+sandboxing was unusable due to its overhead so we want to know if sandboxfs
+makes things better for you.
+
+## Sandboxing basics
+
+Before delving into the details, let's quickly recap what action sandboxing is
+and how sandboxfs is supposed to benefit it. (Refer to [the previous post]({%
+post_url 2017-08-25-introducing-sandboxfs %}) for a much more detailed
+explanation.)
+
+**Sandboxing** is Bazel's ability to isolate the execution of each build action
+(think "compiler invocation") from the rest of the system. This feature
+restricts what actions can do so that they only have access to the tools and
+inputs they declare and so that they are only able to write the outputs they
+promised. In this way, we ensure that the build graph doesn't have hidden
+dependencies that could poison the reproducibility of the build.
+
+More specifically, Bazel constructs an **execroot** for each action, which acts
+as the action's work directory at execution time. The execroot contains all
+input files to the action and serves as the container for any generated outputs.
+Bazel then uses an operating system-provided technique—containers on Linux and
+`sandbox-exec` on macOS—to constrain the action within the execroot. It is worth
+noting that the preparation of the disk layout and the actual sandboxing are
+orthogonal.
+
+Traditionally, Bazel has been creating the execroot using symlinks, thus
+creation scales linearly with the number of inputs. Creating all these symlinks
+is costly for actions with thousands of inputs and unfortunately these are not
+uncommon. This is where sandboxfs is supposed to help.
+
+**sandboxfs** is a FUSE file system that exposes an arbitrary view of the
+underlying file system and does so in constant time. Bazel can then use
```

Comment:
"Constant time" sounds a little awkward. I would say "...underlying file system without time penalties" or "...underlying file system with fixed-time read/write operations."

Reply:
I don't know... "constant time" is well-known terminology and is exactly what I mean here: the time to create the view does not depend on the number of files exposed in the view.

I've taken your "underlying file system without time penalties" suggestion though because the follow-up sentence clarifies what this means anyway.

Reply:
OK if that's standard terminology (I wasn't aware!) then let's leave it as-is.